xonsh/parsers/base.py

```diff
@@ -19,13 +22,55 @@
 from xonsh.tokenize import SearchPath, StringPrefix
 from xonsh.lazyasd import LazyObject
 from xonsh.parsers.context_check import check_contexts
+from xonsh.events import events
+from xonsh.tools import print_exception
 
 RE_SEARCHPATH = LazyObject(lambda: re.compile(SearchPath), globals(),
                            'RE_SEARCHPATH')
 RE_STRINGPREFIX = LazyObject(lambda: re.compile(StringPrefix), globals(),
                              'RE_STRINGPREFIX')
 
 
+events.doc('on_transform_command', """
+on_command_transform(cmd: str) -> str
+
+Fired to request xontribs to transform a command line. Return the transformed
+command, or the same command if no transformaiton occurs.
+
+This may be fired multiple times per command, so design any handlers for this
+carefully.
+""")
+
+
+def transform_command(src, show_diff=True):
```

Comment:
I don't know that the parser is the best place for this, since it is specifically for things beyond the parser. Maybe a new module in the base xonsh package would be better.